crates/goose/src/agents/agent.rs

```diff
@@ -42,32 +43,31 @@ const ESTIMATE_FACTOR_DECAY: f32 = 0.9;
 
 /// The main goose Agent
 pub struct Agent {
-    pub(super) provider: Arc<dyn Provider>,
+    pub(super) provider: Mutex<Option<Arc<dyn Provider>>>,
     pub(super) extension_manager: Mutex<ExtensionManager>,
-    pub(super) frontend_tools: HashMap<String, FrontendTool>,
-    pub(super) frontend_instructions: Option<String>,
-    pub(super) prompt_manager: PromptManager,
-    pub(super) token_counter: TokenCounter,
+    pub(super) frontend_tools: Mutex<HashMap<String, FrontendTool>>,
+    pub(super) frontend_instructions: Mutex<Option<String>>,
+    pub(super) prompt_manager: Mutex<PromptManager>,
+    pub(super) token_counter: Mutex<Option<TokenCounter>>,
     pub(super) confirmation_tx: mpsc::Sender<(String, PermissionConfirmation)>,
     pub(super) confirmation_rx: Mutex<mpsc::Receiver<(String, PermissionConfirmation)>>,
     pub(super) tool_result_tx: mpsc::Sender<(String, ToolResult<Vec<Content>>)>,
     pub(super) tool_result_rx: ToolResultReceiver,
 }
 
 impl Agent {
-    pub fn new(provider: Arc<dyn Provider>) -> Self {
-        let token_counter = TokenCounter::new(provider.get_model_config().tokenizer_name());
+    pub fn new() -> Self {
         // Create channels with buffer size 32 (adjust if needed)
         let (confirm_tx, confirm_rx) = mpsc::channel(32);
         let (tool_tx, tool_rx) = mpsc::channel(32);
 
         Self {
-            provider,
+            provider: Mutex::new(None),
```

Comment:
is decoupling the agent and provider here to support swapping providers during runtime?
for the most part the agent can't function without a provider so having an an option rather than making it required is unexpected

Reply:
looking at the UI code, we get the provider and model info into appConfig:
https://github.com/block/goose/blob/428f0313865fa75ef308282fe07380ec6875321c/ui/desktop/src/main.ts#L261-L275

which seems to be available when we start `goosed` here: 
https://github.com/block/goose/blob/428f0313865fa75ef308282fe07380ec6875321c/ui/desktop/src/main.ts#L331